### PR TITLE
Integrate iOS simulator with device discovery

### PR DESCRIPTION
**Pull request by Fatme**

### CODE REVIEW

**Comment by justcodebuilduser** [comment]
:broken_heart:

**Comment by justcodebuilduser** [comment]
:heart:

File: mobile/mobile-core/devices-service.ts [hunk 1]
```diff
@@ -167,10 +180,12 @@ export class DevicesService implements Mobile.IDevicesService {
 		}).future<void>()();
 	}
 
-	private executeOnAllConnectedDevices(action: (dev: Mobile.IDevice) => IFuture<void>, canExecute?: (_dev: Mobile.IDevice) => boolean): IFuture<void> {
+	private executeOnAllConnectedDevices(action: (dev: Mobile.IDevice) => IFuture<void>, canExecute?: (_dev: Mobile.IDevice) => boolean, options?: {allowNoDevices?: boolean}): IFuture<void> {
```
Comment (rosen-vladimirov) [inline mobile/mobile-core/devices-service.ts]: why have you added the `options` param - I do not see using it anywhere

File: mobile/ios/simulator/ios-emulator-services.ts [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+class IosEmulatorServices implements Mobile.IiOSSimulatorService {
+	constructor(private $logger: ILogger,
+		private $emulatorSettingsService: Mobile.IEmulatorSettingsService,
+		private $errors: IErrors,
+		private $childProcess: IChildProcess,
+		private $devicePlatformsConstants: Mobile.IDevicePlatformsConstants,
+		private $hostInfo: IHostInfo,
+		private $options: ICommonOptions,
+		private $fs: IFileSystem,
```
Comment (teobugslayer) [inline mobile/ios/simulator/ios-emulator-services.ts]: Is $fs used?

File: mobile/mobile-core/ios-simulator-discovery.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
+///<reference path="./../../.d.ts"/>
+"use strict";
+
+import {DeviceDiscovery} from "./device-discovery";
+import Future = require("fibers/future");
+import {IOSSimulator} from "./../ios/simulator/ios-simulator-device";
+
+class IOSSimlatorDiscovery extends DeviceDiscovery {
+	private cachedSimulator: any;
```
Comment (teobugslayer) [inline mobile/mobile-core/ios-simulator-discovery.ts]: Perhaps any -> Mobile.IiSimDevice ?

**Comment by rosen-vladimirov** [comment]
you should update `deviceEmitter` as well. Currently Proton cannot find ios simulators :(

File: bootstrap.ts [hunk 4]
```diff
@@ -73,8 +73,9 @@ $injector.require("deviceLost", "./mobile/mobile-core/device-discovery");
 $injector.require("iTunesValidator", "./validators/iTunes-validator");
 $injector.require("deviceDiscovery", "./mobile/mobile-core/device-discovery");
 $injector.require("iOSDeviceDiscovery", "./mobile/mobile-core/ios-device-discovery");
+$injector.require("iOSSimlatorDiscovery", "./mobile/mobile-core/ios-simulator-discovery");
```
Comment (rosen-vladimirov) [inline bootstrap.ts]: this should be iosSim**u**latorDiscovery

**Comment by Fatme** [comment]
I've updated `deviceEmitter` class

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by justcodebuilduser** [comment]
:heart:

File: mobile/android/android-emulator-services.ts [hunk 5]
```diff
@@ -320,22 +334,18 @@ class AndroidEmulatorServices implements Mobile.IEmulatorPlatformServices {
...
-			} else {
-				_.each(outputRaw, (device: string) => {
-					let rx = device.match(AndroidEmulatorServices.RUNNING_ANDROID_EMULATOR_REGEX);
-
-					if(rx && rx[1]) {
-						emulatorDevices.push(rx[1]);
-					}
-				});
-			}
+			let emulatorDevices = this.$options.geny ? this.getRunningGenymotionEmulators(outputRaw).wait() : this.getRunningAvdEmulators(outputRaw).wait();
```
Comment (teobugslayer) [inline mobile/android/android-emulator-services.ts]: Not a merge stopper, but I'd suggest to rewrite this as an if() else .. else { throw /assert } to help us when we add Visual Studio emulator.

File: mobile/ios/simulator/ios-emulator-services.ts [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+	}
+
+	public checkDependencies(): IFuture<void> {
+		return Future.fromResult();
+	}
+
+	public checkAvailability(dependsOnProject: boolean = true): IFuture<void> {
+		return (() => {
+			if(!this.$hostInfo.isDarwin) {
+				this.$errors.fail("iOS Simulator is available only on Mac OS X.");
```
Comment (teobugslayer) [inline mobile/ios/simulator/ios-emulator-services.ts]: failWithoutHelp, perhaps?

File: mobile/ios/simulator/ios-emulator-services.ts [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+			}
+
+			let platform = this.$devicePlatformsConstants.iOS;
+			if(dependsOnProject && !this.$emulatorSettingsService.canStart(platform).wait()) {
+				this.$errors.fail("The current project does not target iOS and cannot be run in the iOS Simulator.");
```
Comment (teobugslayer) [inline mobile/ios/simulator/ios-emulator-services.ts]: should we use .failWithoutHelp instead?

File: mobile/ios/simulator/ios-simulator-device.ts [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+		return this.$injector.resolve(fileSystemPath.IOSSimulatorFileSystem, { iosSim: this.iosSim, identifier: this.simulator.id });
+	}
+
+	public openDeviceLogStream(): void {
+		return this.iosSim.printDeviceLog(this.deviceInfo.identifier);
+	}
+
+	private _iosSim: any = null;
+	private get iosSim(): any {
+		if (!this._iosSim) {
```
Comment (teobugslayer) [inline mobile/ios/simulator/ios-simulator-device.ts]: require always returns the same object, so you can simplify this code to
```
private get iosSim(): any {
  return require("ios-sim-portable");
}
```

File: mobile/mobile-core/ios-simulator-discovery.ts [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+				return output.indexOf('launchd_sim') !== -1;
+			} catch(e) {
+				return false;
+			}
+		}).future<boolean>()();
+	}
+
+	private _iOSSim: any = null;
+	private get iOSSim(): any {
+		if (!this._iOSSim) {
```
Comment (teobugslayer) [inline mobile/mobile-core/ios-simulator-discovery.ts]: require always returns the same object, so you can simplify this code to
```
private get iosSim(): any {
  return require("ios-sim-portable");
}
```

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by justcodebuilduser** [comment]
:heart:

File: appbuilder/proton-bootstrap.ts [hunk 10]
```diff
@@ -32,6 +33,14 @@ $injector.register("deviceAppDataProvider", {
 		}
 	}
 );
+$injector.register("emulatorSettingsService", {
+	canStart(platform: string): IFuture<boolean> {
+		return Future.fromResult(true);
+	},
+	minVersion(): number {
+		return 17;
```
Comment (rosen-vladimirov) [inline appbuilder/proton-bootstrap.ts]: hm, why 17?

**Comment by justcodebuilduser** [comment]
:broken_heart:

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by justcodebuilduser** [comment]
:broken_heart:

**Comment by justcodebuilduser** [comment]
:broken_heart:

**Comment by justcodebuilduser** [comment]
:broken_heart:

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by justcodebuilduser** [comment]
:broken_heart:

**Comment by justcodebuilduser** [comment]
:heart:

File: mobile/ios/simulator/ios-simulator-device.ts [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+			type: "Emulator"
+		};
+	}
+
+	public get isEmulator(): boolean {
+		return true;
+	}
+
+	public get applicationManager(): Mobile.IDeviceApplicationManager {
+		return this.$injector.resolve(applicationManagerPath.IOSSimulatorApplicationManager, { iosSim: this.$iOSSimResolver.iOSSim, identifier: this.simulator.id });
```
Comment (rosen-vladimirov) [inline mobile/ios/simulator/ios-simulator-device.ts]: cache the object and use it when it already exists :)

File: mobile/ios/simulator/ios-simulator-file-system.ts [hunk 12]
```diff
@@ -0,0 +1,52 @@
+///<reference path="../../../.d.ts"/>
+"use strict";
+
+import Future = require("fibers/future");
+import * as path from "path";
+import * as shell from "shelljs";
```
Comment (rosen-vladimirov) [inline mobile/ios/simulator/ios-simulator-file-system.ts]: `shelljs` is better name :dart:

File: mobile/mobile-core/devices-service.ts [hunk 13]
```diff
@@ -251,6 +294,10 @@ export class DevicesService implements Mobile.IDevicesService {
 					}
 				}
 			}
+
+			if (this.$hostInfo.isWindows && this._platform && this.$mobileHelper.isiOSPlatform(this._platform) && this.$options.emulator) {
```
Comment (rosen-vladimirov) [inline mobile/mobile-core/devices-service.ts]:
```TypeScript
if(!this.$hostInfo.isDarwin...
```

Comment (rosen-vladimirov) [inline mobile/mobile-core/devices-service.ts]:
```TypeScript
if(!this.$hostInfo.isDarwin...
```

File: mobile/mobile-core/devices-service.ts [hunk 14]
```diff
@@ -192,21 +228,27 @@ export class DevicesService implements Mobile.IDevicesService {
...
 				}
+				this.executeCore(action, canExecute).wait();
 			} else {
 				let message = constants.ERROR_NO_DEVICES;
-				if(options && options["allowNoDevices"]) {
+				if (options && options.allowNoDevices) {
 					this.$logger.info(message);
 				} else {
-					this.$errors.failWithoutHelp(message);
+					if (this.$hostInfo.isWindows && this._platform && this.$mobileHelper.isiOSPlatform(this._platform)) {
```
Comment (rosen-vladimirov) [inline mobile/mobile-core/devices-service.ts]:
```TypeScript
if(!this.$hostInfo.isDarwin...
```

File: services/livesync-service-base.ts [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+				_.each(filesToDelete, filePath => shell.rm("-rf", path.join(tempDir, path.relative(projectFilesPath, filePath))));
+
+				// Rename platform specific files
+				let filesToRename = _.filter(localToDevicePaths, ldp => (<any>ldp).onDeviceFileName !== path.basename(ldp.getLocalPath()));
+				_.each(filesToRename, ldp => {
+					let relativePath =  path.relative(projectFilesPath, ldp.getLocalPath());
+					let oldPath = path.join(tempDir, relativePath);
+					let newPath = path.join(tempDir, path.dirname(relativePath), (<any>ldp).onDeviceFileName);
+					if (this.$fs.exists(oldPath).wait()) {
+						this.$fs.rename(oldPath, newPath).wait();
```
Comment (rosen-vladimirov) [inline services/livesync-service-base.ts]: `shelljs.mv("-f")`

File: services/livesync-service-base.ts [hunk 16]
```diff
@@ -0,0 +1,243 @@
...
+			this.$logger.info("Successfully transferred all project files.");
+		}).future<void>()();
+	}
+
+	private resolvePlatformLiveSyncService(platform: string, device: Mobile.IDevice): IPlatformLiveSyncService {
+		return this.$injector.resolve(this.$liveSyncProvider.platformSpecificLiveSyncServices[platform.toLowerCase()], {_device: device});
+	}
+
+	private getCanExecuteAction(platform: string, appIdentifier: string): (dev: Mobile.IDevice) => boolean {
+		let canExecute: (dev: Mobile.IDevice) => boolean = null;
```
Comment (rosen-vladimirov) [inline services/livesync-service-base.ts]:
```TypeScript
let canExecute: (dev: Mobile.IDevice) => dev.applicationManager.isApplicationInstalled(appIdentifier).wait()
```

File: services/project-files-manager.ts [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+			filePath: filePath,
+			onDeviceFileName: path.basename(filePath),
+			shouldIncludeFile: true
+		};
+	}
+
+	private parseFile(filePath: string, validValues: string[], value: string): any {
+		let regex = util.format("^(.+?)[.](%s)([.].+?)$", validValues.join("|"));
+		let parsed = filePath.match(new RegExp(regex, "i"));
+		if (parsed) {
```
Comment (rosen-vladimirov) [inline services/project-files-manager.ts]:
```TypeScript
return parsed ||  {
 				filePath: filePath,
 				onDeviceFileName: path.basename(parsed[1] + parsed[3]),
 				shouldIncludeFile: parsed[2].toLowerCase() === value.toLowerCase(),
 				value: value
 			}
```

**Comment by justcodebuilduser** [comment]
:heart:

File: services/livesync-service-base.ts (hunk 16 above)
Comment (Fatme) [inline services/livesync-service-base.ts]: This is not valid in case when the application is not installed on device.

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by rosen-vladimirov** [comment]
:+1:  great work

**Pull request merged**